### DISCUSSION TITLE
The icon order in the Systray widget changes on every Qtile restart

### DISCUSSION THREAD

**Discussion by s-aleck** [qtile/qtile#3195] [Q&A] [answered]
When I restart Qtile, the icon order in Systray changes. For example, the Flameshot icon will no longer be the first one, but third. Can this be fixed?
(Using Qtile in Arco Linux.)

**Comment by elParaguayo** [comment, marked as answer]
Are you restarting or just reloading the config? If you're restarting, I suspect there's not much that can be done but, if you're reloading the config, maybe we can store the order somewhere.

**Reply by s-aleck** [reply]
I'm restarting Qtile by pressing a key combination (Meta + Shift + R). I didn't check what happens when I restart Arco.

**Reply by elParaguayo** [reply]
Thanks. What I meant was Qtile has a `lazy.restart()` call and a `lazy.reload_config()` one. `restart` starts a new qtile process whereas `reload_config` just reloads the config in the same process. I assume your keybinding is to `restart`.

If you're just restarting because of changes you've made to your config then try `reload_config` instead. I suspect the icon order will still change but there might be something we can do to fix that.

When you restart, I think we're at the mercy of the order that your applications register themselves with the tray.

**Reply by s-aleck** [reply]
Oh. Sorry. I went back to my `config.py`. It was `restart`. I switched it to `reload_config`. I notice this is a bit faster than restarting, but the icon order  changes every time I reload the configuration.

**Reply by elParaguayo** [reply]
No problem at all.

I thought the order would still change as it's not determined by the widget, it's determined by the applications registering themselves with the tray. There are no rules for how the icons should be ordered so it's "first come, first served" at the moment!

I can take a look at seeing if it's possible to memorise the order between reloads.

There's some other work that needs to be done to the Systray too so, when I get round to doing that, I'll take a look at this at the same time.

**Reply by elParaguayo** [reply]
Hmm. From a quick look, I'm not sure how possible this is. When the config is reloaded, the applications request new icons and so there's a new window ID for the icons after each load. This means that, even if we can save the previous icon IDs, they won't match the new ones.

**Reply by m-col** [reply]
We could do something like alphabetise the icons so they are always in a deterministic order. We wouldn't need to persist any state across reloads/restarts.

**Reply by elParaguayo** [reply]
That's what I'm looking at now!

**Reply by elParaguayo** [reply]
Done: https://github.com/qtile/qtile/pull/3198
